### DISCUSSION TITLE
Server Sent Events (SSE)

### DISCUSSION THREAD

**Discussion by RothAndrew** [basdijkstra/rest-assured-net#125] [Q&A] [answered]
Is it possible to use rest-assured-net to test an endpoint that does Server Sent Events (SSE)?

I have an SSE endpoint I'm trying to test now, and the behavior that I'm seeing is that the GET request does not complete until the server closes out the connection. In code terms, I'm seeing rest-assured-net behave the same way that I would expect the following code snippet to behave:

```
httpClient.SendAsync(request, HttpCompletionOption.ResponseContentRead, cancellationToken)
```

with the important bit being `ResponseContentRead` which instructs HttpClient to wait until the server has sent all content

As opposed to

```
httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
```

which will return the 200 and the headers right away and let you stream the content using `response.Content.ReadAsStreamAsync()`.

Is it possible to tell rest-assured-net to use `ResponseHeadersRead` so that I can test that my endpoint is returning the right status code and headers? Reading the content as a stream and being able to validate stuff in there would be cool too but not absolutely necessary for what I want to test.

Here's an example test from my API project that I'm working with that interfaces with Argo Workflows. The `logstream` endpoint returns Newline-Delimited Json (NDJSON) using the SSE pattern.

```
[Fact]
public void GetWorkflowLogstream_WhenCalledOnRunningPod_ReturnsOkImmediately()
{
    // Submit the workflow
    var workflow = (WorkflowInfo)Given()
        .Accept(MediaTypeNames.Application.Json)
        .ContentType(MediaTypeNames.Application.Json)
        .Body(new WorkflowSubmission()
        {
            TemplateName = TemplateName,
            GenerateName = GenerateName,
            Parameters = [new WorkflowParameter("waitSeconds", "5")]
        })
        .When()
        .Post($"{RootUrl}/api/v1/workflows/")
        .Then()
        .StatusCode(HttpStatusCode.OK)
        .And()
        .DeserializeTo(typeof(WorkflowInfo));
    Thread.Sleep(500);

    // loop until the workflow is running
    Stopwatch sw = Stopwatch.StartNew();
    while (workflow.Status != WorkflowStatus.Running && sw.ElapsedMilliseconds < 3000)
    {
        workflow = (WorkflowInfo)Given()
            .Log(RequestLogLevel.All)
            .Accept(MediaTypeNames.Application.Json)
            .ContentType(MediaTypeNames.Application.Json)
            .When()
            .Get($"{RootUrl}/api/v1/workflows/{workflow.Name}")
            .Then()
            .StatusCode(HttpStatusCode.OK)
            .And()
            .DeserializeTo(typeof(WorkflowInfo));
        Thread.Sleep(500);
    }
    sw.Stop();

    // Get the logstream. Make sure it isn't waiting for the workflow to finish
    Given()
        .Log(RequestLogLevel.All)
        .Timeout(TimeSpan.FromSeconds(1))
        .Accept("application/x-ndjson")
        .When()
        .Get($"{RootUrl}/api/v1/workflows/{workflow.Name}/logstream")
        .Then()
        .Log(ResponseLogLevel.All)
        .StatusCode(HttpStatusCode.OK)
        .And()
        .ContentType("application/x-ndjson")
        .And()
        .Header("Cache-Control", "no-cache")
        .And()
        .Header("Connection", "keep-alive")
        .And()
        .ResponseTime(NHamcrest.Is.LessThan(TimeSpan.FromSeconds(1)));
}
```

**Comment by basdijkstra** [comment, marked as answer]
Hey @RothAndrew, thank you for your message.

Would it be enough for you if I would provide a method in the DSL, for the RequestSpecification and a flag in the config to make RestAssured use `ResponseHeadersRead` instead of `ResponseContentRead`?

You’re correct that it always uses the latter at the moment because that’s the default for `SendAsync()`: https://github.com/basdijkstra/rest-assured-net/blob/9a5c71fdc8d26940503deec4fa20f6baf129acf4/RestAssured.Net/Request/HttpRequestProcessor.cs#L131

**Reply by RothAndrew** [reply]
I believe it would 😄

**Reply by basdijkstra** [reply]
Only one way to find out then!

I think this is a pretty minor change, the only problem is I don’t know how to write tests for it yet (I use WireMock.Net and I don’t know how to trigger server side events there).

What I’ll do is implement it as discussed, create and publish a beta release so you can use that to test it. Would that work for you?

I can’t do that before next week, though, as I’m still on holiday at the moment. Back to work on Monday.

**Reply by RothAndrew** [reply]
Sounds good! And no worries at all! Appreciate all your work on this project, it has been really nice to work with.

**Comment by basdijkstra** [comment]
Tracking progress here: https://github.com/basdijkstra/rest-assured-net/issues/126
